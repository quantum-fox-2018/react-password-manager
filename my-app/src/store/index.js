import {createStore, applyMiddleware, combineReducers} from 'redux'
import thunk from 'redux-thunk'
import passwordReducers from './password/password.reducers'
import userReducers from './users/user.reducers'
import logger from 'redux-logger'


const reducers = combineReducers({
  passManager : passwordReducers,
  user: userReducers
})

const store = createStore(
  reducers,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
  applyMiddleware(thunk, logger)
)

export default store