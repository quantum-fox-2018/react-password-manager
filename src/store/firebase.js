import firebase from 'firebase'

const config = {
  apiKey: "AIzaSyCJ0Hlgag0DOOKpxzUsGTaE4FAu0tIE_EE",
  authDomain: "react-password-manager-bef30.firebaseapp.com",
  databaseURL: "https://react-password-manager-bef30.firebaseio.com",
  projectId: "react-password-manager-bef30",
  storageBucket: "react-password-manager-bef30.appspot.com",
  messagingSenderId: "827996324966"
}
firebase.initializeApp(config)
export const Database = firebase.database()
export const User = firebase.auth()