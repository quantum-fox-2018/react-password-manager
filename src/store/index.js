import { createStore, applyMiddleware, combineReducers } from "redux"
import logger from 'redux-logger'
import usersReducers from './users/reducers'
import thunk from 'redux-thunk'
import passmanagerReducer from "./passmanager/reducers";

const reducers = combineReducers({
  users: usersReducers,
  passmanager: passmanagerReducer
})

let store = createStore(reducers, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(), applyMiddleware(thunk, logger))

export default store