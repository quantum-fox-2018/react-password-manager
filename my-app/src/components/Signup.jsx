import React, { Component } from 'react';
import bcrypt from 'bcryptjs'
import {db} from '../firebase'
import firebase from 'firebase'
import jwt from 'jsonwebtoken'
const saltRounds = 10;
const salt = bcrypt.genSaltSync(saltRounds);

class Signup extends Component {
  constructor () {
    super ()
    this.state ={
      username: '',
      password: ''
    }
  }

  handleChange = (e) => {
    // e.preventDefault()
    this.setState({
      [e.target.name] : e.target.value
    })
  }

  addNewUser = () => {
    let {history} = this.props
    let hash = bcrypt.hashSync(this.state.password,salt)
    let newUser = {
      username: this.state.username,
      password: hash,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      updatedAt: firebase.database.ServerValue.TIMESTAMP
    }
    // this.props.signupUser(newUser)
    db.ref('/pass-user').push(newUser)
    .then(() => {
      let token = jwt.sign({username: newUser.username,password: newUser.password}, 'kitten')
      localStorage.setItem('username', this.state.username)
      localStorage.setItem('token', token)
      history.push('/home')
    })
  }

  lowerCase () {
    return (this.state.password.match(/^(?=.*[a-z])/))
  }
  
  upperCase () {
    return (this.state.password.match(/^(?=.*[A-Z])/)) 
  }
  
  oneNum () {
    return (this.state.password.match(/^(?=.*[0-9])/))
  }
  
  minLength () {
    return (this.state.password.length > 6) ? true : false
  }

  specialChar () {
    return (this.state.password.match(/^(?=.*[_\W])/))
  }

  allValid() {
    return (this.minLength() && this.upperCase() && this.lowerCase() && this.oneNum() && this.specialChar())
  }
  render() {
    return (
      <div className="container">
        <h1>Sign up page</h1>
        <form>
          <div className="form-group">
            <label>Username</label>
            <input type="text" id="username" name="username" className="form-control"
              value={this.state.username}
              onChange={this.handleChange}
            />
          </div>
          <div className="form-group">
            <label>Password</label>
            <input type="password" name="password" className="form-control"
              value={this.state.password}
              onChange={this.handleChange}
            />
          </div>
          <button type="button" className={"btn btn-primary "+ (this.allValid() ? "active" : "disabled")} onClick={this.addNewUser}>Sign up</button>
        </form>
        <div style={{marginTop:'10px'}}>
          <h4>Password requirements: </h4>
            <ul className="text-left" style={{listStyle: "none"}}>
              <li className={"alert alert-dismissible " + (this.lowerCase() ? "alert-success" : "alert-danger")}> {this.lowerCase() ? '[ O ]': '[ X ]'} At least <strong>one lowercase</strong></li>
              <li className={"alert alert-dismissible " + (this.upperCase() ? "alert-success" : "alert-danger")}> {this.upperCase() ? '[ O ]': '[ X ]'} At least <strong>one capital</strong></li>
              <li className={"alert alert-dismissible " + (this.oneNum() ? "alert-success" : "alert-danger")}> {this.oneNum() ? '[ O]': '[ X ]'} At least <strong>one number</strong></li>
              <li className={"alert alert-dismissible " + (this.minLength() ? "alert-success" : "alert-danger")}> {this.minLength() ? '[ O ]': '[ X ]'} Minimum <strong>6 characters</strong></li>
              <li className={"alert alert-dismissible " + (this.specialChar() ? "alert-success" : "alert-danger")}> {this.specialChar() ? '[ O ]': '[ X ]'} At least <strong>Use one symbol</strong></li>
            </ul>
          </div>
      </div>
    );
  }
}

export default Signup;