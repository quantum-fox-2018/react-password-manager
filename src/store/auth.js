import { observable } from 'mobx'
import swal from 'sweetalert2'
import { auth } from './firebase'


class Auth {
  userId = ''
  email = ''
  @observable statusAuth = false

  errorNotif(err) {
    swal({
      position: 'center',
      type: 'error',
      title: 'Check your password',
      text: err.message
    })
  }

  successNotif(info) {
    swal({
      position: 'center',
      type: 'success',
      title: info,
      showConfirmButton: false,
      timer: 1000
    })
  }

  signUp(email, password) {
    auth.createUserWithEmailAndPassword(email, password)
    .then(res => {
      this.email = email
      this.userId = res.uid
      this.statusAuth = true
      this.successNotif('Enjoy!')
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }

  signIn(email, password) {
    auth.signInWithEmailAndPassword(email, password)
    .then(res => {
      this.email = email
      this.userId = res.uid
      this.statusAuth = true
      this.successNotif('Welcome Back!')
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }

  signOut() {
    this.email = ''
    this.userId = ''
    auth.signOut()
    .then(res => {
      this.successNotif('See You!')
      this.statusAuth = false
    })
    .catch(err => {
      this.errorNotif(err)
    })
  }
}

export default new Auth()
