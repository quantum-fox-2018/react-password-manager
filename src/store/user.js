import { observable } from 'mobx'
import swal from 'sweetalert2'


import { db, auth } from './firebase'
import authData from './auth'

class PassManager {
  @observable listData = []
  valUrl = false
  valLength = false
  valNumber = false
  valUpper = false
  valLower = false
  valSpecial = false
  allVal = false

  checkUrl(url) {
    let regex = /[-a-zA-Z0-9@:%_+.~#?&//=]{2,256}\.[a-z]{2,4}\b(\/[-a-zA-Z0-9@:%_+.~#?&//=]*)?/gi
    this.valUrl = regex.test(url)
    return this.valUrl
  }

  checkLength(password) {
    this.valLength = (password.length >= 6)
    return this.valLength
  }

  checkNumber(password) {
    this.valNumber = (/\d/g.test(password))
    return this.valNumber
  }

  checkUpper(password) {
    this.valUpper = (/[A-Z]/g.test(password))
    return this.valUpper
  }

  checkLower(password) {
    this.valLower = (/[a-z]/g.test(password))
    return this.valLower
  }

  checkSpecial(password) {
    this.valSpecial = (/\W/g.test(password))
    return this.valSpecial
  }

  checkAllValidation() {
    if(this.valUrl && this.valLength && this.valNumber && this.valUpper && this.valLower && this.valSpecial) {
      this.allVal = true
      return this.allVal
    }
  }

  hashPassword(str) {
    let length = str.length
    let bullet = String.fromCharCode(0x2022)
    return bullet.repeat(length)
  }

  showPassword(userId, idData) {
    swal.mixin({
      input: 'password',
      confirmButtonText: 'Confirm',
      showCancelButton: true,
    }).queue([
      {
        title: 'Password',
        text: 'enter your password'
      }
    ]).then((result) => {
      if (!result.value) {
        swal({
          type: 'question',
        })
      } else if (result.value[0]) {
        auth.signInWithEmailAndPassword(authData.email, result.value[0])
        .then(res => {
          let data = this.listData.map(data => {
            if(data.id === idData) {
              return {...data, showPassword: !data.showPassword}
            } else {
              return {...data}
            }
          })
          this.listData = data  
        })
        .catch(err => {
          swal({
            type: 'error',
            title: 'Error',
            text: err.message
          })
        })
      }
    })
  }

  loadData(userId) {
    db.ref(`/user/${userId}`).once('value', (snap) => {
      const pass = snap.val()
      let arr = []
      for (const key in pass) {
        if (pass.hasOwnProperty(key)) {
          const el = pass[key]
          const update = {...el, id: key, showPassword: false}
          arr.push(update)
        }
      }
      this.listData = arr
    })
  }

  createData(userId, url, username, password) {
    db.ref(`/user/${userId}`).push({
      url, username, password, createdAt: Date.now(), updatedAt: null
    })
    .then(res => {
      swal({
        position: 'center',
        type: 'success',
        title: 'Your password has been saved',
        showConfirmButton: false,
        timer: 1500
      })
      this.loadData(userId)
    })
    .catch(err => {
      alert(err)
    })
  }

  deleteData(userId, idData) {
    swal({
      title: 'Are you sure?',
      text: "You won't be able to revert this!",
      type: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#3085d6',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Yes, delete it!'
    }).then((result) => {
      if (result.value) {
        db.ref(`/user/${userId}`).child(idData).remove()
        this.loadData(userId)
        swal(
          'Deleted!',
          'Your file has been deleted.',
          'success'
        )
      }
    })
  }

  editPassword(userId, idData) {
    swal.mixin({
      input: 'password',
      confirmButtonText: 'Confirm',
      showCancelButton: true,
      inputAttributes: {
        'minlength': 6,
        'autocapitalize': 'off',
        'autocorrect': 'off'
      }
    }).queue([
      {
        title: 'Edit Password',
        text: 'enter your new password'
      }
    ]).then((result) => {
      if (result.value) {
        let data = this.listData.map(data => {
          if(data.id === idData) {
            let newData = {...data, password: result.value[0], updatedAt: Date.now()}
            db.ref(`/user/${userId}`).child(idData).update(newData)
            return newData
          } else {
            return {...data}
          }
        })
        this.listData = data
        swal({
          position: 'center',
          type: 'success',
          title: 'Your password has been edited',
          showConfirmButton: false,
          timer: 1500
        })
      } else {
        swal({
          type: 'question',
        })
      }
    })
  }
}

export default new PassManager()
