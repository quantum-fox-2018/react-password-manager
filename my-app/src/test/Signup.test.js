import React from 'react'
import Enzyme, {shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import store from '../store'
import LocalStorageMock from './localstorageMock'
import Signup from '../components/Signup'
import {Provider} from 'react-redux'
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'

global.localStorage = new LocalStorageMock();
Enzyme.configure({ adapter: new Adapter() });

describe('<Signup/>', () => {
  it('it should have username and password start with empty list', () => {
    const wrapper = shallow(<Signup/>)
    expect(wrapper.state('username')).toBe('')
    expect(wrapper.state('password')).toBe('')
  })

  it('renders without crashing', () => {
    const div = document.createElement('div');
    const wrapper = shallow(
    <Signup/>
    )
    ReactDOM.render(wrapper, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('it should have form tag ', () => {
    const wrapper = shallow(<Signup/>)
    expect(wrapper.find('form'))
    expect(wrapper.exists('<input type="text">')).toBe(true)
    expect(wrapper.exists('<input type="password">')).toBe(true)
  })

  it('onChange, username state should change', () => {
    const wrapper = shallow(<Signup/>)
    wrapper.find('#username').simulate('change', 
    {target: {name: 'username', value:'testing'}})
    expect(wrapper.state('username')).toBe('testing')
  })
  it('renders without crashing snapshot', () => {
      const rendered = renderer.create(<Signup/>)
      expect(rendered).toMatchSnapshot()
    });
})