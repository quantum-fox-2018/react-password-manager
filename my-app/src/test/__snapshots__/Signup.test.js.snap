// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders without crashing snapshot 1`] = `
<div
  className="container"
>
  <h1>
    Sign up page
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label>
        Username
      </label>
      <input
        className="form-control"
        id="username"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label>
        Password
      </label>
      <input
        className="form-control"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary disabled"
      onClick={[Function]}
      type="button"
    >
      Sign up
    </button>
  </form>
  <div
    style={
      Object {
        "marginTop": "10px",
      }
    }
  >
    <h4>
      Password requirements: 
    </h4>
    <ul
      className="text-left"
      style={
        Object {
          "listStyle": "none",
        }
      }
    >
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one lowercase
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one capital
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          one number
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         Minimum 
        <strong>
          6 characters
        </strong>
      </li>
      <li
        className="alert alert-dismissible alert-danger"
      >
         
        [ X ]
         At least 
        <strong>
          Use one symbol
        </strong>
      </li>
    </ul>
  </div>
</div>
`;
